### DISCUSSION TITLE
In JSON Template Layout, can you handle messages differently based on their type?

### DISCUSSION THREAD

**Discussion by SubieGold** [apache/logging-log4j2#3122] [Q&A]
I posted a StackOverflow, but forgot about this discussion model in GitHub.
https://stackoverflow.com/questions/79120207/how-do-you-log-a-message-if-your-event-template-expects-a-map-but-doesnt-receiv

--- Copying Message to here ---

I want to be able to use Log4j2 to log in JSON format. I am using JsonTemplateLayout and my own Event Template. My application uses MapMessage when logging so I can extract out values for my log format.

The problem is that not all logs are coming from my source code, some could be from Spring or other libraries that may not be sending a MapMessage.

The problem is that when these libraries are logging without MapMessage my JSON EventTemplate doesn't extract out the message to log.

Full Code Available: https://github.com/LoganDev99/log4j2map

Java Logging Snippet - Log with MapMessage

    public static void main(String[] args) {
        SpringApplication.run(Log4j2mapApplication.class, args);

        log.info(new MapMessage<>()
                .with("METHOD", "main()")
                .with("MESSAGE", "My Map Message Event")
        );
    }
log4j2.xml - Using JsonTemplateLayout and additivity false.

```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration>

    <Appenders>
        <Console name="CONSOLE" target="SYSTEM_OUT">
            <JsonTemplateLayout eventTemplateUri="classpath:MyLayout.json"/>
        </Console>
        <Console name="DEFAULT_CONSOLE" target="SYSTEM_OUT">
            <JsonTemplateLayout />
        </Console>
    </Appenders>

    <Loggers>
        <Logger name="com.example" level="INFO" additivity="false">
            <AppenderRef ref="CONSOLE" />
        </Logger>
        <Root level="WARN">
            <AppenderRef ref="DEFAULT_CONSOLE"/>
        </Root>
    </Loggers>

</Configuration>
```

MyLayout.json (Event Template) - Using map resolver

```
{
  "instant": {
    "$resolver": "timestamp",
    "pattern": {
      "format": "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
      "timeZone": "UTC"
    }
  },
  "message": {
    "$resolver": "map",
    "key": "MESSAGE"
  },
  "method": {
    "$resolver": "map",
    "key": "METHOD"
  }
}
```

Console Output

```
{"instant":"2024-10-24T02:06:02.322Z"}
{"instant":"2024-10-24T02:06:02.330Z"}
{"instant":"2024-10-24T02:06:03.461Z"}
{"instant":"2024-10-24T02:06:03.465Z","message":"My Map Message Event","method":"main()"}
```

Notice how in the first 3 logs there's is no message because in this case it's Spring logging. Spring logs such as "Started Log4j2mapApplication in 1.3 seconds" for example.

So my question is how do have a fallback to just log the message if the message isn't a map.

**Comment by vy** [comment]
@LoganDev99, I think what you're looking for is `fallbackKey` of [the `message` resolver](https://logging.staged.apache.org/log4j/2.x/manual/json-template-layout.html#event-template-resolver-message). Given the following Log4j configuration

```xml
<JsonTemplateLayout eventTemplate='{"$resolver": "message", "fallbackKey": "literal"}'/>
```

below Java code

```java
LOGGER.info("plain text");
LOGGER.info(new MapMessage<>().with("key", "value"));
```

outputs

```
{"literal":"plain text"}
{"key":"value"}
```

Does this help?

**Reply by SubieGold** [reply]
Not quite, 

If I use that approach the map would be logged. But if I was also extracting out a key then now I have duplicate in the log. See how `main()` appears twice.

```
{"instant":"2024-10-25T18:50:12.107Z","log.logger":"com.example.log4j2map.Log4j2mapApplication","message":{"MESSAGE":"My Map Message Event","METHOD":"main()"},"method":"main()"}

```

Sample template
```
  "message": {"$resolver": "message", "fallbackKey": "literal"},
  "method": {
    "$resolver": "map",
    "key": "METHOD"
  }
```

**Reply by vy** [reply]
> Not quite,
> 
> If I use that approach the map would be logged. But if I was also extracting out a key then now I have duplicate in the log. See how `main()` appears twice.
> 
> ```json
> {
>   "instant":"2024-10-25T18:50:12.107Z",
>   "log.logger":"com.example.log4j2map.Log4j2mapApplication",
>   "message":{
>     "MESSAGE":"My Map Message Event",
>     "METHOD":"main()"
>   },
>   "method":"main()"
> }
> ```

@LoganDev99, why do you need the custom `method` property given that the complex `message` object contains it already? Also note that you have a complex `message` object, and a property referring to a particular attribute of that object; what output are you expecting without duplication?

**Reply by SubieGold** [reply]
Method was just an example. I have other stuff in the map that I'm logging. 

So I ideally the log would look like this.

Log Line 1: Is a log **not** providing a `MapMessage` from Spring.
Log Line 2: Is a log from my application providing `MapMessage`.

So what I want is to be able to use a map. But have a fallback to `message` if the event isn't a `Map`. It's similar to the fallbackKey suggestion, but on a map instead of message.

```
{"instant":"2024-10-28T16:02:06.755Z","message":"Started Log4j2mapApplication in 1.336 seconds (process running for 1.978)"}
{"instant":"2024-10-28T16:02:06.758Z","message":"My Map Message Event","method":"main()"}
```

**Comment by vy** [comment]
@LoganDev99, how do you expect a `MapMessage` to generate the `"My Map Message Event"` string? Your `message` property in the 2nd JSON you shared should be an object, right?

**Reply by SubieGold** [reply]
My original template is in the original post. The log I'm posting is desired outcome. It's not happening today.


```
{
  "instant": {
    "$resolver": "timestamp",
    "pattern": {
      "format": "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
      "timeZone": "UTC"
    }
  },
  "message": {
    "$resolver": "map",
    "key": "MESSAGE"
  },
  "method": {
    "$resolver": "map",
    "key": "METHOD"
  }
}
```

**Reply by vy** [reply]
Thanks for the clarification, I see. I guess you're looking for some sort of `isMapMessage ? message["message"] : message`. JTL doesn't provide this. I guess for these kind of  needs we might introduce a generic script-based resolver, i.e., one that users can provide a script to generate the output.

Curious: What is your use case necessitating an individual `.method` property instead of using `.message.METHOD`? (Assuming `{"$resolver": "message", "fallbackKey": "literal"}` is used.)

**Reply by SubieGold** [reply]
Yup, I think you got it.

> Curious: What is your use case necessitating an individual .method property instead of using .message.METHOD? (Assuming {"$resolver": "message", "fallbackKey": "literal"} is used.)

Maybe my usage of MESSAGE as keys in the map may have cause confusion. But technically I would be ok using the `flatten` config. It's just going back to my original request, not all message are maps.

```
  "message": {
    "$resolver": "map",
    "flatten": true
  }
```
